pkg/storage/filter.go

```diff
@@ -0,0 +1,143 @@
+package storage
+
+import (
+	"fmt"
+	"sort"
+
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// A FilterExpression describes an AST for record stream filters.
+type FilterExpression interface {
+	isFilterExpression()
+}
+
+// FilterExpressionFromStruct creates a FilterExpression from a protobuf struct.
+func FilterExpressionFromStruct(s *structpb.Struct) (FilterExpression, error) {
+	if s == nil {
+		return nil, nil
+	}
+	return filterExpressionFromStruct(nil, s)
+}
+
+func filterExpressionFromValue(path []string, v *structpb.Value) (FilterExpression, error) {
+	switch vv := v.GetKind().(type) {
+	case *structpb.Value_ListValue:
+		var or OrFilterExpression
+		for _, vvv := range vv.ListValue.Values {
+			e, err := filterExpressionFromValue(path, vvv)
+			if err != nil {
+				return nil, err
+			}
+			or = append(or, e)
+		}
+		return or, nil
+	case *structpb.Value_StructValue:
+		return filterExpressionFromStruct(path, vv.StructValue)
+	}
+	return filterExpressionFromEq(path, v)
+}
+
+func filterExpressionFromStruct(path []string, s *structpb.Struct) (FilterExpression, error) {
+	var and AndFilterExpression
+	var fs []string
+	for f := range s.GetFields() {
+		fs = append(fs, f)
+	}
+	sort.Strings(fs)
+
+	for _, f := range fs {
+		v := s.GetFields()[f]
+		switch f {
+		case "$and":
+			expr, err := filterExpressionFromValue(path, v)
+			if err != nil {
+				return nil, err
+			}
+			or, ok := expr.(OrFilterExpression)
```

Comment:
should it be `AndFilterExpression` here?

Reply:
yes, good catch

Reply:
nevermind, it was an `or`. By default we're treating a list as an `or`. The `$and` transforms that to an `and`.